packages/flutter_tools/lib/src/preview_device.dart

```diff
@@ -0,0 +1,185 @@
+
+BundleBuilder _defaultBundleBuilder() {
+  return BundleBuilder();
+}
+
+class PreviewDevice extends Device {
+  PreviewDevice({
+    @required Platform platform,
+    @required ProcessManager processManager,
+    @required Logger logger,
+    @required FileSystem fileSystem,
+    @visibleForTesting BundleBuilderFactory builderFactory = _defaultBundleBuilder,
+  }) : _platform = platform,
+       _processManager = processManager,
+       _logger = logger,
+       _fileSystem = fileSystem,
+       _bundleBuilderFactory = builderFactory,
+       super('preview', ephemeral: false, category: Category.desktop, platformType: PlatformType.custom);
+
+  final Platform _platform;
+  final ProcessManager _processManager;
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  final BundleBuilderFactory _bundleBuilderFactory;
+
+  @override
+  void clearLogs() { }
+
+  @override
+  Future<void> dispose() async { }
+
+  @override
+  Future<String> get emulatorId async => null;
+
+  final DesktopLogReader _logReader = DesktopLogReader();
+
+  @override
+  FutureOr<DeviceLogReader> getLogReader({covariant ApplicationPackage app, bool includePastLogs = false}) => _logReader;
+
+  @override
+  Future<bool> installApp(covariant ApplicationPackage app, {String userIdentifier}) async => true;
+
+  @override
+  Future<bool> isAppInstalled(covariant ApplicationPackage app, {String userIdentifier}) async => false;
+
+  @override
+  Future<bool> isLatestBuildInstalled(covariant ApplicationPackage app) async => false;
+
+  @override
+  Future<bool> get isLocalEmulator async => false;
+
+  @override
+  bool isSupported() => true;
+
+  @override
+  bool isSupportedForProject(FlutterProject flutterProject) => true;
+
+  @override
+  String get name => 'preview';
+
+  @override
+  DevicePortForwarder get portForwarder => const NoOpDevicePortForwarder();
+
+  @override
+  Future<String> get sdkNameAndVersion async => 'preview';
+
+  Process _process;
+  Directory _assetDirectory;
+
+  @override
+  Future<LaunchResult> startApp(covariant ApplicationPackage package, {
+    String mainPath,
+    String route,
+    @required DebuggingOptions debuggingOptions,
+    Map<String, dynamic> platformArgs,
+    bool prebuiltApplication = false,
+    bool ipv6 = false,
+    String userIdentifier,
+  }) async {
+    _assetDirectory = _fileSystem.systemTempDirectory
+      .createTempSync('flutter_preview.');
+
+    // Build assets and perform initial compilation.
+    Status status;
+    try {
+      status = _logger.startProgress('Compiling application for preview...');
+      await _bundleBuilderFactory().build(
+        buildInfo: debuggingOptions.buildInfo,
+        mainPath: mainPath,
+        platform: TargetPlatform.tester,
+        assetDirPath: getAssetBuildDirectory(),
+      );
+      copyDirectory(_fileSystem.directory(
+        getAssetBuildDirectory()),
+      _assetDirectory.childDirectory('data').childDirectory('flutter_assets'),
```

Comment:
I think the whitespace is wrong but hard to tell in GitHub
```suggestion
        _assetDirectory.childDirectory('data').childDirectory('flutter_assets'),
```